src/components/ArtworkDetailPage/ButtonGroup.tsx

```diff
@@ -0,0 +1,102 @@
+'use client';
+
+import { useEffect, useState } from 'react';
+import { useStore } from 'zustand';
+
+import Icon from '@/components/Icon/Icon';
+import CollectionModal from '@/components/Modal/CollectionModal';
+import useToggleLike from '@/hooks/serverStateHooks/useToggleLike';
+import modalStore from '@/stores/modalStore';
+import { ArtworkPostSocialInfoType } from '@/types';
+
+import ShareModal from '../Modal/ShareModal';
+
+interface ButtonGroupProps {
+  socialInfo: ArtworkPostSocialInfoType;
+}
+
+const ButtonGroup = ({
+  socialInfo: { postId, likeCount: initialLikeCount, isLiked: initialIsLiked },
+}: ButtonGroupProps) => {
+  const { openModal } = useStore(modalStore);
+  const [blockButton, setBlockButton] = useState(false);
+
+  const {
+    mutate: toggleLike,
+    setLikeStatus,
+    isLiked,
+    likeCount,
+  } = useToggleLike({
+    isLiked: initialIsLiked,
+    likeCount: initialLikeCount,
+  });
+
+  const openCollectionModal = () => {
+    openModal({
+      modalSize: 'lg',
+      contents: <CollectionModal />,
+    });
+  };
+
+  const openShareModal = () => {
+    openModal({
+      modalSize: 'md',
+      contents: <ShareModal />,
+    });
+  };
+
+  const handleClickLikeToggle = () => {
+    setBlockButton(true);
+    toggleLike(postId);
+    setBlockButton(false);
+  };
```

Comment:
```suggestion
  const handleClickLikeToggle = async () => {
    setBlockButton(true);
    await toggleLike(postId); // 비동기 작업 대기
    setBlockButton(false);
  };
```

이렇게 비동기 작업을 대기하도록 하는 건 어떻게 생각하시나여?

Reply:
좋습니다..! mutate가 promise 객체를 반환하고 있지 않아서 별도 비동기 처리가 필요없다고 생각하고 있었습니다..!

혹시 다른 mutate를 실행하는 함수에서는 별도의 async/await을 하지 않던데 해당 부분에서만 추가되는 특별한 이유가 있을까요? setLBlockButton을 함수 전후로 활용하고 있기 때문일까요?

Reply:
네, `setBlockButton`이 `toggleLike` 함수의 앞뒤로 있기 때문이 맞습니다!

`async/await`를 추가하길 제안드린 이유는 '유튜브 댓글 좋아요처럼 작동할 것을 염려해서'였습니다. 비동기 작업을 대기하지 않고 낙관적 업데이트를 진행하면 유튜브처럼 댓글 좋아요가 반영되지 않았음에도 불구하고 일단 UI 먼저 바꾸게 되면 발생할 문제가 있어서용

아직 서버에 좋아요가 반영되지 않았는데 UI는 반영된 것으로 보여서 사용자가 페이지를 이탈하거나 다른 클릭을 시도할 경우 (제가 자주 그럽니다), 좋아요 반영하는 함수 실행이 중단되기 때문에 UX상으로 불편하리라 생각합니다! 물론 유튜브는 좋아요 수가 워낙 많아서 느린 거겠지만 저는 불편했습니다!

'`mutate`를 실행하는 함수에서는 별도의 `async/await`을 하지 않는다'고 하셨는데 이 부분은 `async/await`를 따로 개발자가 작성하지 않아도 내부적으로 실행될 코드가 동기적으로 작동하는 것이 맞는지 확인해 볼 필요가 있을 것 같아용

저는 이렇게 생각합니다 ㅎㅎ 상우님은 어떻게 생각하시나요?

Reply:
위에서 제가 말씀드렸던 것 처럼 matate자체가 promise 객체를 반환하지 않기 때문에 우리가 실제로 사용하는 `mutateFn` 비동기 로직이더리도 내부적으로 처리를 하기 때문에 실제 결과만 받아 사용하겠지라는 추측에서 시작하게 되었습니다..!

그리고 동기적으로 동작하는 toggleLike의 결과에 따라 onSuccess에서 좋아요 수를 변화를 적용하기 때문에 짧은 시간에 버튼이 연속으로 눌리는 동작만 막는 것만 추가하면 되겠다고 판단하였습니다..!

[이 링크에 참조 된 글](https://velog.io/@leitmotif/mutate%EC%99%80-mutateAsync%EC%9D%98-%EB%8B%A4%EB%A5%B8-%EC%A0%90%EC%97%90-%EB%8C%80%ED%95%B4)은 mutate관련한 tanstack-query 내부로직 흐름을 적어 두신 블로그리 공유 드립니다..!


낙관적 업데이트에 대해서 저는 약간 다른 생각을 가지고 있습니다.
좋아요 / 팔로우는 오히려 낙관적 업데이트를 사용해야한다고 생각하는 편입니다.

좋아요 팔로우가 사용자가 요청에 대한 결과가 로직 처리 후 반영 된다면, 클릭 후 상태 변경 까지 대기 시간은 오히려 동작하지 않는 것 처럼 느낄 것 같습니다. (로딩 UI 처리를 한다면 또 다르겠지만). 그리고 저는 좋아요/팔로우 기능은 요청에 대한 응답을 기대하기 보다 내가 누른 행동에 대한 피드백을 더 빨리 받아 볼 수 있을 때 더 좋은 기능이라고 생각해서 낙관적 업데이트를 적극적으로 활용하는 편이었습니다!

그리고 말씀해주신 특정 행동에 대해서 함수의 실행이 중단될 수 있으니 UX 상으로 불편할 것 같다는 점은 좋아요/팔로우의 경우 프론트 딴 로직은 요청을 보내고, 상태를 바꾸어주는 것 외는 크게 필요한 기능이 없다고 생각합니다. 말씀해주신 함수의 실행은 브라우저에서 보내는 요청을 받은 서버에서 일어나는 처리 과정에 해당하는 단계라고 추측이 되는데, 브라우저의 창 닫힘이나 다른 행동에 의해서 좋아요를 처리하는 서버 로직이 중단되는 것이 이상하다는 생각도 듭니다.! 


( 약간 생각을 쓰면서 두서없이 쓴 것 같아 이해하기 어려운 부분이 있을 것 같습니다..! 있다면 알려주세용 ㅎㅎ )

Reply:
오 그런 이유에서 하신 거였군요.
'좋아요/팔로우 기능은 요청에 대한 응답을 기대하기 보다 내가 누른 행동에 대한 피드백을 더 빨리 받아 볼 수 있을 때 더 좋은 기능'이라는 점도 납득했습니다!

제가 말씀드린 유튜브 댓글 좋아요 이슈는 확실히 백엔드의 문제가 맞는 것 같아요. 백엔드가 좋아요 데이터를 처리하지 않았기 때문에 좋아요 UX가 안 좋아진 것 같습니다.

좋은 의견 감사합니다! 👍🏻✨
`TanStack Query`와 백엔드 공부를 더 했다면 이 부분도 좀 더 깊이 고민할 수 있었을 것 같아요. 참고 자료랑 명쾌한 답변 정말 감사합니다!! 해당 로직 수정 안 하고 그대로 진행하셔도 좋습니당